### ISSUE TITLE
Canvas2D: exceptions during animation

### ISSUE THREAD

**Issue by WhiteHexagon** [mirkosertic/Bytecoder#367]
I am seeing some strange index out of bounds errors whilst animating my scene, but maybe only after around 80,000 iterations.  I use something similar to your Kotlin example:

```
    animationCallback = AnimationFrameCallback { aElapsedTime ->
        //my drawing code
        window.requestAnimationFrame(animationCallback)
    }
    window.requestAnimationFrame(animationCallback)
```

At first I had it wrapped in a separate class to simplify the calling code and encapsulate some fps timing logic.  But now I have stripped everything out, and with just a single `println("loop")` statement as my drawing logic, I get exceptions after ~80,000 frames.

So I am thinking it could be either the recursion?, or possible something garbage collection based?  So I am now just trying to break the recursion pattern to rule that out.

However I can not find support for things like Thread.sleep,wait,notify, or the Concurrent package, or any other mechanism to avoid the recursion...  

Sorry to say that I am still using June'19 build because of the UTF-8 issue I have with the new build.  So apologies if this has been fixed.  

I tried leaving your 2D ball demo running, and it works fine in this version of firefox.

btw I added Binaryen `wasm-opt` to my build chain, and your hand generated wasm seems very efficient!  since it barely shrank the wasm by much at all :)

**Comment by WhiteHexagon** [comment]
Hmm, now I am wondering if the call to `requestAnimationFrame` is actually asynchronous...  the fact that it is using a callback too.  And the above usage pattern looks correct according to the info I could find.

[https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame](https://developer.mozilla.org/en-US/docs/Web/API/Window/requestAnimationFrame)

It doesn't solve my problem, but this is probably nothing to do with stack overflows / recursion.

**Comment by mirkosertic** [comment]
`window.requestAnimationFrame` is by default async. There is no need to add thread synchronization here. The exception you are encountering is most likely related to a garbage collector bug which was fixed in later release. Please upgrade to the most recent version of Bytecoder, as I cannot provide bug fix releases for older versions.

**Comment by WhiteHexagon** [comment]
Ok good to know thanks!  I am very eager to try the new LLVM backend, but I still have the exception mentioned in #333 

`Error linking class sun.nio.cs.UTF_8`

Using:

`Kotlin version 1.3.21-release-158 (JRE 1.8.0_161-b12)`

**Comment by WhiteHexagon** [comment]
OK I just tried 2020-04-09, now with working llvm backend :)  but I am seeing some jitter/stalling on the canvas 2D animation performance.  It might be because of the way I have the new garbage collector, or something else.  I try to investigate more.

**Comment by WhiteHexagon** [comment]
I tried to add some frames-per-second measurement, since it seems to degrade over time as well.  However I seem to be getting a negative number for `System.currentTimeMillis()` or maybe it is the println statement in Kotlin.  e.g.

`println("frames = $frames  ms ${System.currentTimeMillis()}")`

prints:

`frames = 1  ms -1495642666`

and in case it was the println, I tried this:

```
val now = System.currentTimeMillis()
            if (now < 0) {
                println("negative")
            }
```
which does print negative.

P.S. The value is increasing over time (ie moving towards positive)

**Comment by mirkosertic** [comment]
System.currentTimeMillis() is not a good way for performance measurements. The Bytecoder Wasm backend cannot represent Long datatypes, as it is limited to the Wasm32 datatypes. The current unix timestamp does not fit into the 32 bit datatypes, hence the overflow resultung in a negative value. Even the JS backend cannot represent Longs correctly due to lack of emulation and the JavaScript 53bit precision floating point type. I'm waiting for BigInt and Wasm64 support available in all browsers to fix this.

Maybe the W3C `User Timing API` can help here, or maybe the Chrome Autiding tools in the meanwhile,, but as mentioned, the profiling support for WebAssembly is pretty limited here.

**Comment by WhiteHexagon** [comment]
Interesting, thanks.  I am surprised though, I thought the wasm32 vs wasm64 was more about the memory addressing range?  But I thought Wasm itself had a 64 bit int type.  But I am only just learning about all this, so maybe it is a browser limitation?  I tried to find some more information, but apart from the Wasm spec listing i64, I didn't find much information.

[https://webassembly.github.io/spec/core/syntax/types.html](https://webassembly.github.io/spec/core/syntax/types.html)

**Comment by mirkosertic** [comment]
Wasm-32 was part of the MVP. Maybe the 64 bit datatypes are now supported by some browsers, but they are useless as they cannot transported to JS side without a conversion and performance impact. Wasm-BigInt integration is discussed as a Post-MVP Feature, as seen here:  https://webassembly.org/docs/future-features/

**Comment by WhiteHexagon** [comment]
Yes I understand now, clearly there is no good solution :(  BigInt also does not sound ideal.  However I think i64 was part of the MVP from what I have read today.

So just to be clear then,  all js `number` types coming from javascript code will be truncated to Wasm `int32` type?  or maybe an f32??  or is f64 also supported there?

I tried adding the following js:

```
        bytecoder.imports.extmisc.now = function (thisref) {
            return new Date().getTime();
        }; 
```

and I see that this is also truncated.  I had a look through your code to try and find if there was something similar to `toBytecoderReference`.  I was thinking maybe there are some `toFloat64` or `toLong` type methods, but I didnt see anything. 

I also spent a couple hours now scratching my head :)  I thought I would be clever and pass my js 'number' (53bit int) as a length 2 `de.mirkosertic.bytecoder.api.web.IntArray` to my Kotlin, and reassemble on my side as a Kotlin `Long` ... but my '<< shift left'  `lsh` wraps at 32 bits :(

So does that mean bytecoder is not supporting Long at all?

**Comment by WhiteHexagon** [comment]
PS wasn't linux/unix always 32bit for 'epoch' time stamps?  I seem to remember reading recently that they only just added 64bit time support to linux.  So shouldn't the current time fit into a 32 bit signed number?  It has been a long day, so sorry if I misunderstand something.

**Comment by WhiteHexagon** [comment]
update: the stutter on the animation seems to be related to the 'new' garbage collector.  So that would explain why the June'19 build is so smooth!  that was before the GC was added?  So I only just noticed I can change the GC period, and at 2 seconds I get silky smooth animation, then a big pause for the GC.  So of course this now opens up new questions.  
1. can the GC be disabled?
2. Is there a way to working with bytecoder that no garbage is generated?  
3. Is there a way to detect if I am leaking objects, since I realised now that my 'star constellation' tile, that I stop after 300 frames, is actually running out of memory when it crashes (so maybe it was not recursion or a bug in the old builds...)

My preference would be to find a way to generate no garbage, since anyway Wasm does not yet have a GC.

**Comment by mirkosertic** [comment]
The garbage collector can be disabled, but you shouldn't, because you will for sure run into out-of-memory problems pretty soon. Just remove the setTimeout invocation initialiting the GC invocation on JS side.

It is pretty hard to write JVM/Java/Kotlin code that does not put load onto the garbage collector. You have to:

* avoid usage of the `new` keyword, in consequence everything must be `static` to be reachable
* reuse arrays
* avoid `automatic in/outboxing` of JVM primitives, as this creates a lot of short-time living object.s
* respect the above mentioned rules twice as much in the "hot" paths of your code

The GC is a key ingredient of the JVM and any managed language running on it. The Bytecoder GC is currently not very sophisticated, so there is room for improvement.

**Comment by WhiteHexagon** [comment]
Thank you for the information.  Before I try and attempt this, is there anything on the bytecoder side that might be also be generating new objects or garbage behind the scenes?

**Comment by mirkosertic** [comment]
Bytecoder has currently some overhead for InvokeDynamic handling, leading to temporary array creation. This can be avoided and is on my todo list. Everything else is related to the JRE classes, which are linked from OpenJDK. So there should't be too much overhead.

Since you are using Kotlin, I suggest to check the following link: https://medium.com/@BladeCoder/exploring-kotlins-hidden-costs-part-2-324a4a50b70

Kotlins type-check intrinsics can be disabled for production builds. So I guess we can reduce some Kotlin overhead by passing the mentioned option to the compiler.

I don't think you have to or should rewrite all your code. I think we should improve the compiler where it is neccesary. Could you please create a debug build for your example, with debugOutput=true and minifyCompileResult=false? Then we can try to use the Chrome profiling tools to see where the runtime goes. The best weapon is a good escape analysis done by the compiler to avoid heap fragmentation. Non-Escaping objects can be put on the stack instead of heap, reducing GC load and reducing memory allocation costs.

**Comment by mirkosertic** [comment]
@WhiteHexagon I've release Bytecoder `2020-05-04` today. It includes some performance improvements in the runtime environment and the garbage collector. Also the `wasm_llvm` backend now fully supports long and double datatypes. If you like you can give it a try.

**Comment by WhiteHexagon** [comment]
Awesome, I am very impressed at the speed you added that, I didn't even get chance to reply to the last comment yet :)  I give it a try right now since I am currently sat working on that code.

**Comment by mirkosertic** [comment]
Yes it takes some time to sync to the mirrors. Give it one day :-)

**Comment by WhiteHexagon** [comment]
got it! I will give it a test tomorrow :)

**Comment by WhiteHexagon** [comment]
Long support seems to be working, so that is great! :)  I tested (`getTime()` on a js `Date`) and the left shift thing I was hacking.  Both work find now :)  Although my IDE switched out the bytecoder IntArray import for kotlins IntArray which had me confused for a while :)

btw I forgot to mention on this version, and the last, I started getting this warning in the console:

```
Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/
main.js:297:21
```

297: refers to `request.open('HEAD',jsPath,false);` in the `unixfilesystem` block.

Performance is difficult to judge, I still see quite a pause after 2 seconds, or stuttering if I use 200ms.  I have to investigate more what Kotlin is doing, since I thought it avoided boxing/unboxing if it could.  Maybe I try and split just one tile out, that I can do the debug build and work out what is going on.  It will be next week though now.

I added a pause button to my animator, but even then I still see a gradual decrease in free memory.  So I need to go back to basics with a simple example, since I now have about 24 tiles in my demo.

**Comment by WhiteHexagon** [comment]
I found a possible problem with one of my `Long` transforms that I was trying to hack the current time with.  It is just a hex string coming over from javascript (and yes I was missing the `toBytecoderString` to start with).  However now I have the hex string arriving on the kotlin side, and I try to parse that into a long:

```
  val n = now()  //javascript call
  println("string = $n")
  val l = n.toLong(16)
  println("long = $l")
```

string = 171f072b1e1
long = 70862057

Kotlin parses this fine as: 1588876980705

[https://pl.kotl.in/i-f9GK83C](https://pl.kotl.in/i-f9GK83C)

This is using latest CLI build and llvm target

**Comment by WhiteHexagon** [comment]
OK, so now I am back to something more basic to work with.  I cleaned up a lot of the code, and removed the very annoying gradle tool and switched back to maven, phew!! :)  So now everything should be self contained with the pom `install` task.  Although my maven skills are a bit rusty now, so apologies for that mess.  Consider this work-in-progress, like most my code ;)

[https://github.com/WhiteHexagon/example-kotlin-bytecoder-wasm](https://github.com/WhiteHexagon/example-kotlin-bytecoder-wasm)

**Comment by WhiteHexagon** [comment]
I realised I could get rid of the `toDouble()` calls by implementing Canvas2D `arc` in my own extension class.  Not sure if it helps, but just trying to simplify the calls.  But I got thinking about how strings are handled...  and I realised my map keys are very inefficient, so I will next convert those to 2D array.

**Comment by WhiteHexagon** [comment]
OK, I worked out the strange syntax for 2D arrays in kotlin :) and performance is looking much better (a lot less string garbage).  Previously it was falling over after a few hundred frames, now it only starts to struggle around 2500 frames.  I pushed my fixes to my repo. above.

**Comment by mirkosertic** [comment]
After some profiling of your example, I found a really nasty and obvious bug in the LLVM backend : #405. This will fix the GC performance issues in the LLVM backend.

**Comment by WhiteHexagon** [comment]
Great work! I looking forward to trying it out, and I am glad the smaller demo was some help in isolating that.

**Comment by mirkosertic** [comment]
A new Bytecoder release is on the way and is syncing to the mirrors

**Comment by WhiteHexagon** [comment]
Sadly I am having some troubles with this version.  The behavior on my 'mini' demo code seems the same.  It still drains memory, and starts to stutter after 2000 or so frames.   

And the 'multi-tile' demo, fails with a `RuntimeError: unreachable executed`.  I found the `enableexceptions` option, but didn't help.  I switched back to the previous 4th May version, and it works, switched back, and the above error.  It will take me some time to comment out the various tiles to try and identify the culprit, but are there any other changes I should be aware of that might be the reason?

**Comment by mirkosertic** [comment]
Well, maybe the new escape anapysis is doig too much causing the RuntimeErrors. I will add an option to this, making it by default off for now.

**Comment by mirkosertic** [comment]
I've relased a new version with escape analysis disabled by default for now. In the meanwhile, i am working on some improvements, for instance #412.

**Comment by WhiteHexagon** [comment]
Great! I tried commenting out various parts of my larger demo, but the results are inconclusive.  Sometimes it complained about a missing Canvas2D extension, which was being used elsewhere, and then sometimes it was the above exception.

To be honest I spent yesterday porting the smaller 'Stars' demo back to 'Go' (which is where I started my WASM experiments) but this time using 'tinygo'...  I wanted to try a language without GC, and didnt fancy messing around with Rust at this point.

I pushed the code to my github, and it shows some advantage of using the single js/web API interface approach (no custom extensions required).  

I also wanted to investigate my wasm binary size.  Previously 'Go' was generating ~1.3MB binary (not LLVM based), my Kotlin version is now at ~500KB (LLVM & wasm-opt), and interesting tinygo (LLVM & wasm-opt) is coming in at 100KB.  I also found and installed 'twiggy' wasm analyser, and I think maybe a large chunk of the 100KB is debug.  I am just trying to see if there is a tool to 'prune' the binaries further now, especially as I imagine there is a lot of the Kotlin standard stuff I am not making use of, and probably plenty of other stuff like filesystem, i18n, coroutines, data/time, unicode, etc.  Have you seen any tools that can do that sort of pruning on the JVM bytecode side, i.e. pre bytecoder? or you think it is better to try and do that once I have my kotlin as WASM?

**Comment by WhiteHexagon** [comment]
I am still working on this, but after I commented out huge chunks of my larger demo code, I had various issues getting it to build again :(  Firstly it seemed like IntelliJ was caching some stuff, then gradle was doing some strange stuff, so I switched also this project back to Maven, but then that seemed also to be generating the .js file based from a cached jar, or possibly from .m2 archive.  So now I am porting the thing back to `ant` so I can really see what is going on.  Mostly it seems like my `abstract class` is not getting processed into the generated .js file (all the extension `interface` based files work), and no idea why at this point.  Slowly trying to eliminate all the possibilities, I keep you posted.  (I am using latest version now btw)

**Comment by WhiteHexagon** [comment]
good news, I am back up and running with 2020-05-15.  I tried the -escapeAnalysis analysis flag, and no errors.  Is there any feedback from that option that I know for sure it did something?  Also, are there any internal bytecoder properties that I can access at runtime that would provide a version string such as `2020-05-20` to ensure things are correctly setup here?

**Comment by mirkosertic** [comment]
Hey! I'd like to know how your project is going, as the due to this pandemic stuff most of the work on Bytecoder needed to be postponed. How are you doing?

**Comment by WhiteHexagon** [comment]
No problem.  Currently I am still experimenting with TinyGo (LLVM), mainly because of smaller wasm binary, and GC performance.  I might even try Zig next, since that seems to have manual memory management :)  I only ported part of my Kotlin code so far, since I also had lots of detours along the way.  I just recently got my WebSocket stuff into a WebWorker which helps with the frame rate.  I would love to come back to Kotlin at some point, since some of this stuff is painful in Go, like pushing memory blocks back and forth between Go&JS, and development by the Go team seems to have stopped.  So keep me posted, and I look forward to seeing what you do next with the project.
